File: app/src/main/java/io/tanjundang/study/knowledge/service/RemoteActivity.java
package io.tanjundang.study.knowledge.service;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;

import io.tanjundang.study.R;

public class RemoteActivity extends AppCompatActivity {

    private RemoteAidlInterface mBinder;
    private ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            mBinder = RemoteAidlInterface.Stub.asInterface(iBinder);
            try {
                mBinder.calcMsg();
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_remote);
    }

    public void StartRemoteService(View v) {
        Intent intent = new Intent(this, RemoteService.class);
        bindService(intent, connection, BIND_AUTO_CREATE);
    }

    public void UnbindRemoteService(View v) {
        if (mBinder != null) {
            unbindService(connection);
            mBinder = null;
        }
    }
}

File: app/src/main/java/io/tanjundang/study/knowledge/service/RemoteActivity.kt
package io.tanjundang.study.knowledge.service

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import android.os.RemoteException
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View

import io.tanjundang.study.R
import io.tanjundang.study.base.BaseActivity

class RemoteActivity : BaseActivity() {

    private var mBinder: RemoteAidlInterface? = null
    private val connection = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            mBinder = RemoteAidlInterface.Stub.asInterface(iBinder)
            try {
                mBinder!!.calcMsg()
            } catch (e: RemoteException) {
                e.printStackTrace()
            }

        }

        override fun onServiceDisconnected(componentName: ComponentName) {

        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun initView() {
        setContentView(R.layout.activity_remote)
    }

    override fun initData() {

    }

    fun StartRemoteService(v: View) {
        val intent = Intent(this, RemoteService::class.java)
        bindService(intent, connection, Context.BIND_AUTO_CREATE)
    }

    fun UnbindRemoteService(v: View) {
        if (mBinder != null) {
            unbindService(connection)
            mBinder = null
        }
    }
}
